crates/amaru-stores/src/rocksdb/mod.rs

```diff
@@ -125,82 +143,119 @@ impl RocksDB {
+        let opts = default_opts_with_prefix();
+        rocksdb::DB::open_for_read_only(&opts, dir.join("live"), false)
+            .map(|db| ROLedgerDB { db })
+            .map_err(|err| StoreError::Internal(err.into()))
     }
 
     pub fn empty(dir: &Path, era_history: &EraHistory) -> Result<RocksDB, StoreError> {
         let mut opts = Options::default();
         opts.create_if_missing(true);
         opts.set_prefix_extractor(SliceTransform::create_fixed_prefix(PREFIX_LEN));
-        Ok(RocksDB {
-            dir: dir.to_path_buf(),
-            incremental_save: true,
-            db: OptimisticTransactionDB::open(&opts, dir.join("live"))
-                .map_err(|err| StoreError::Internal(err.into()))?,
-            era_history: era_history.clone(),
-            ongoing_transaction: OngoingTransaction::new(),
-        })
+        OptimisticTransactionDB::open(&opts, dir.join("live"))
+            .map(|db| Self {
+                dir: dir.to_path_buf(),
+                incremental_save: true,
+                db,
+                era_history: era_history.clone(),
+                ongoing_transaction: OngoingTransaction::new(),
+            })
+            .map_err(|err| StoreError::Internal(err.into()))
     }
 
     fn transaction_ended(&self) {
         self.ongoing_transaction.set(false);
     }
 }
 
+fn decode_cbor<T: for<'d> cbor::decode::Decode<'d, ()>>(
+    bytes: Option<impl AsRef<[u8]>>,
+) -> Result<Option<T>, StoreError> {
+    bytes
+        .map(|b| cbor::decode(b.as_ref()))
+        .transpose()
+        .map_err(StoreError::Undecodable)
+}
+
 fn get<T: for<'d> cbor::decode::Decode<'d, ()>>(
     db: &OptimisticTransactionDB,
     key: &str,
 ) -> Result<Option<T>, StoreError> {
-    db.get(key)
-        .map_err(|err| StoreError::Internal(err.into()))?
-        .map(|bytes| cbor::decode(&bytes))
-        .transpose()
-        .map_err(StoreError::Undecodable)
+    let bytes = db
+        .get(key)
+        .map_err(|err| StoreError::Internal(err.into()))?;
+    decode_cbor(bytes)
 }
 
 #[allow(clippy::panic)]
 #[allow(clippy::unwrap_used)]
-fn iter<'a, K: Clone + for<'d> cbor::Decode<'d, ()>, V: Clone + for<'d> cbor::Decode<'d, ()>>(
-    db: &OptimisticTransactionDB,
+fn iter_decode<'a, K, V, I>(it: I) -> impl Iterator<Item = (K, V)> + 'a
+where
+    K: for<'d> cbor::Decode<'d, ()> + 'a,
+    V: for<'d> cbor::Decode<'d, ()> + 'a,
+    I: Iterator<Item = Result<(Box<[u8]>, Box<[u8]>), rocksdb::Error>> + 'a,
+{
+    it.map(|e| {
+        let (key, value) = e.unwrap();
+        let k = cbor::decode(&key[PREFIX_LEN..])
+            .unwrap_or_else(|e| panic!("unable to decode key ({}): {e:?}", hex::encode(&key)));
+        let v = cbor::decode(&value).unwrap_or_else(|e| {
+            panic!(
+                "unable to decode value ({}) for key ({}): {e:?}",
+                hex::encode(&value),
+                hex::encode(&key)
+            )
+        });
+        (k, v)
+    })
+}
+
+pub fn iter_db<'a, K, V>(
```

Comment:
The iter / iter_db (and same for get) duplication could be removed by introducing some traits that would be implemented by twice. Not yet clear if it's worth the hassle.